src/backend/gpopt/translate/CQueryMutators.cpp

```diff
@@ -35,6 +35,513 @@ extern "C" {
+//
+//	@doc:
+//		Checks if list grouping_list contains all columns from conkeys of
+//		Primary Key in relation conrelid. grouping_list is treated as flat
+//		list of SortGroupClauses, other nodes are ignored.
+//---------------------------------------------------------------------------
+BOOL
+CQueryMutators::GroupingListContainsPrimaryKey(Query *query,
+											   List *grouping_list,
+											   List *conkeys, Oid conrelid)
+{
+	BOOL result = false;
+	ULONG found_col_cnt = 0;
+	ListCell *lc_conkey;
+
+	// go over every conkey of constraint
+	ForEach(lc_conkey, conkeys)
+	{
+		AttrNumber attnum = lfirst_int(lc_conkey);
+		BOOL found_col = false;
+		ListCell *lgc;
+
+		// search for conkey in grouping list
+		ForEach(lgc, grouping_list)
+		{
+			Node *grpcl = (Node *) lfirst(lgc);
+
+			if (NULL == grpcl || !IsA(grpcl, SortGroupClause))
+			{
+				continue;
+			}
+
+			SortGroupClause *sgc = (SortGroupClause *) grpcl;
+			Node *expr = gpdb::GetSortGroupClauseExpr(sgc, query->targetList);
+
+			GPOS_ASSERT(IsA(expr, Var));
+
+			Var *var = (Var *) expr;
+
+			RangeTblEntry *rte =
+				(RangeTblEntry *) gpdb::ListNth(query->rtable, var->varno - 1);
+
+			GPOS_ASSERT(NULL != rte);
+
+			if (var->varattno == attnum && var->varlevelsup == 0 &&
+				rte->relid == conrelid)
+			{
+				found_col = true;
+				found_col_cnt++;
+				break;
+			}
+		}
+		if (!found_col)
+		{
+			break;
+		}
+	}
+
+	// check that all conkeys were found
+	if (gpdb::ListLength(conkeys) == found_col_cnt)
+	{
+		result = true;
+	}
+
+	return result;
+}
+
+//---------------------------------------------------------------------------
+//	@function:
+//		CQueryMutators::AddMissingGroupClauseMutator
+//
+//	@doc:
+//		Traverses the groupClause, adds new SortGroupClause (defined by m_gc
+//		field in context) to all grouping lists where m_gc has functional
+//		dependency.
+//---------------------------------------------------------------------------
+Node *
+CQueryMutators::AddMissingGroupClauseMutator(
+	Node *node, SContextFixGroupDependentTargets *context)
+{
+	if (NULL == node)
+	{
+		return NULL;
+	}
+
+	if (IsA(node, SortGroupClause))
+	{
+		if (context->m_parent_is_grouping_clause)
+		{
+			// In case the SortGroupClause is direct element of groupsets list
+			// of GroupingClause, we should treat it in a special way - consider
+			// it like a 1-element list.
+			List *new_list = ListMake1((Node *) gpdb::CopyObject(node));
+			if (GroupingListContainsPrimaryKey(context->m_query, new_list,
+											   context->m_conkeys,
+											   context->m_conrelid))
+			{
+				new_list = gpdb::LAppend(
+					new_list, gpdb::CopyObject((Node *) context->m_gc));
+				return (Node *) new_list;
+			}
+			gpdb::ListFreeDeep(new_list);
+		}
+		return (Node *) gpdb::CopyObject(node);
+	}
+
+	if (IsA(node, GroupingClause))
+	{
+		GroupingClause *original_grouping_clause = (GroupingClause *) node;
+		GroupingClause *new_grouping_clause = makeNode(GroupingClause);
```

Comment:
I think you should make a wrapper for makeNode too.

Reply:
There is already one existing - ```MakeNode```

Reply:
Then use it